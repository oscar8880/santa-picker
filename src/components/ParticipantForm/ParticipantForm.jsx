import styles from './ParticipantForm.module.scss';
import React from 'react';
import FormikTextInput from '../../components/FormikTextInput/FormikTextInput';
import { Row, Col } from 'react-grid-system';
import * as Yup from 'yup';
import Text from '../../components/Text/Text';
import SantaHatIcon from '../Icons/SantaHatIcon/SantaHatIcon';
import {
  includesCurrencySymbol,
  validCurrencyAmount,
} from '../../constants/regex';

export const getFormValidationSchema = (
  numParticipants = 3,
  spendingLimitChecked = false,
) => {
  const schema = {};
  for (let i = 0; i < numParticipants; i++) {
    schema[`participantName${i}`] = Yup.string()
      .required('Required')
      .min(2, 'Must be at least two characters')
      .max(30, 'Must be 30 characters or fewer');
    schema[`participantEmail${i}`] = Yup.string()
      .email('Invalid email address')
      .required('Required')
      .max(100, 'Must be 100 characters or fewer');
  }

  if (spendingLimitChecked) {
    schema.spendingLimit = Yup.string()
      .required('Required')
      .matches(includesCurrencySymbol, 'Please specify a currency')
      .matches(validCurrencyAmount, 'Please enter a valid amount');
  }

  schema.organiserName = Yup.string()
    .required('Required')
    .min(2, 'Must be at least two characters')
    .max(30, 'Must be 30 characters or fewer');

  return Yup.object().shape(schema);
};

export const getInitialValues = (
  numParticipants = 3,
  spendingLimitChecked = false,
) => {
  const values = {};
  for (let i = 0; i < numParticipants; i++) {
    values[`participantName${i}`] = '';
    values[`participantEmail${i}`] = '';
  }

  if (spendingLimitChecked) {
    values.spendingLimit = '';
  }

  values.organiserName = '';

  return values;
};

const iconColours = [
  '#FF3A2F',
  '#FFA500',
  '#FFCB17',
  '#798517',
  '#226F61',
  '#AD85FF',
  '#CA61C3',
];

export const generateNameAndEmailFields = (numParticipants = 3) => {
  const fields = [];
  for (let i = 0; i < numParticipants; i++) {
    fields.push(
      <div key={`name-email-row-${i}`}>
        <Row>
          <Col>
            <div className={styles.ParticipantForm__Participant}>
              <div
                className={styles.ParticipantForm__Icon}
                style={{
                  backgroundColor: iconColours[i % 7],
                }}
              >
                <SantaHatIcon width={25} />
              </div>
              <Text appearance="h5">{`Participant ${i + 1}`}</Text>
            </div>
          </Col>
        </Row>
        <Row key={`participantKey${i}`}>
          <div className={styles.ParticipantForm__InputRow}>
            <Col xs={12} sm={6}>
              <div className={styles.ParticipantForm__Input}>
                <FormikTextInput
                  label="Name"
                  name={`participantName${i}`}
                  id={`participantName${i}`}
                  type="text"
                />
              </div>
            </Col>
            <Col xs={12} sm={6}>
              <div className={styles.ParticipantForm__Input}>
                <FormikTextInput
                  label="Email"
                  name={`participantEmail${i}`}
                  id={`participantEmail${i}`}
                  type="text"
                />
              </div>
            </Col>
          </div>
        </Row>
      </div>,
    );
  }

  return fields;
};
