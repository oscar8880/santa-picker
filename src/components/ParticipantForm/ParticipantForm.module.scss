@import "../../scss/functions";
@import "../../scss/variables";

.ParticipantForm__InputRow {
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  width: 100%;
}

.ParticipantForm__Input {
  padding-bottom: $spacer-6;
  @media (min-width: $breakpoint-md) {
    padding-bottom: $spacer-8;
  }
}

.ParticipantForm__Participant {
  align-items: center;
  display: flex;
  flex-direction: row;
  justify-content: center;
  margin-bottom: $spacer-2;
  @media (min-width: $breakpoint-md) {
    justify-content: left;
  }
}

.ParticipantForm__Icon {
  align-items: center;
  border-radius: 50%;
  display: flex;
  height: 40px;
  justify-content: center;
  margin-right: $spacer-2;
  opacity: 0.7;
  width: 40px;
}
