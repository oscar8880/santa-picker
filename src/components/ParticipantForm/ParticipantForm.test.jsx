import React from 'react';
import { render } from '@testing-library/react';
import { Formik, Form } from 'formik';

import { generateNameAndEmailFields } from './ParticipantForm';

test('matches snapshot', async () => {
  const { asFragment } = render(
    <Formik initialValues={{}} onSubmit={() => {}}>
      <Form>{generateNameAndEmailFields()}</Form>
    </Formik>,
  );

  expect(asFragment()).toMatchSnapshot();
});
