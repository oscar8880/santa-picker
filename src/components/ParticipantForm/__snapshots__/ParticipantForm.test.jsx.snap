// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <div>
      <div
        style="margin-left: -15px; margin-right: -15px; display: flex; flex-wrap: wrap; flex-grow: 0; flex-shrink: 0; align-items: normal; justify-content: flex-start;"
      >
        <div
          style="box-sizing: border-box; min-height: 1px; position: relative; padding-left: 15px; padding-right: 15px; width: 100%; flex-basis: 0px; flex-grow: 1; flex-shrink: 0; max-width: 100%; margin-left: 0%;"
        >
          <div
            class="ParticipantForm__Participant"
          >
            <div
              class="ParticipantForm__Icon"
              style="background-color: rgb(255, 58, 47);"
            >
              <svg
                viewBox="0 0 512 512"
                width="25"
              >
                <path
                  class="prefix__active-path"
                  d="M421.522 360.98c-41.359 0-75.01 33.652-75.01 75.01 0 41.359 33.652 76.01 75.01 76.01 41.359 0 75.01-34.652 75.01-76.01 0-41.359-33.651-75.01-75.01-75.01zM405.011 62.517C365.338 22.844 311.596 0 255.5 0S146.662 22.844 106.989 62.517s-61.517 93.415-61.517 149.511v58.939h331.045v-75.01c0-8.292 6.71-15.002 15.002-15.002s15.002 6.71 15.002 15.002v136.534c4.943-.716 9.865-1.516 15.002-1.516 16.167 0 31.311 3.977 45.006 10.533v-129.48c-.001-56.096-21.845-109.837-61.518-149.511z"
                  data-old_color="#000000"
                  data-original="#000000"
                  fill="#fff"
                />
                <path
                  class="prefix__active-path"
                  d="M45.472 300.972c-16.571 0-30.004 13.433-30.004 30.004v30.004c0 16.571 13.433 30.004 30.004 30.004h281.57a105.212 105.212 0 0149.475-49.475v-40.537H45.472z"
                  data-old_color="#000000"
                  data-original="#000000"
                  fill="#fff"
                />
              </svg>
            </div>
            <p
              class="Text Text__h5"
            >
              Participant 1
            </p>
          </div>
        </div>
      </div>
      <div
        style="margin-left: -15px; margin-right: -15px; display: flex; flex-wrap: wrap; flex-grow: 0; flex-shrink: 0; align-items: normal; justify-content: flex-start;"
      >
        <div
          class="ParticipantForm__InputRow"
        >
          <div
            style="box-sizing: border-box; min-height: 1px; position: relative; padding-left: 15px; padding-right: 15px; width: 50%; flex-basis: 50%; flex-grow: 0; flex-shrink: 0; max-width: 50%; margin-left: 0%;"
          >
            <div
              class="ParticipantForm__Input"
            >
              <div
                class="FormikTextInput__Container"
              >
                <label
                  class="FormikTextInput__Label FormikTextInput__LabelAsPlaceholder"
                  for="participantName0"
                >
                  Name
                </label>
                <input
                  class="FormikTextInput__Input"
                  id="participantName0"
                  name="participantName0"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
          <div
            style="box-sizing: border-box; min-height: 1px; position: relative; padding-left: 15px; padding-right: 15px; width: 50%; flex-basis: 50%; flex-grow: 0; flex-shrink: 0; max-width: 50%; margin-left: 0%;"
          >
            <div
              class="ParticipantForm__Input"
            >
              <div
                class="FormikTextInput__Container"
              >
                <label
                  class="FormikTextInput__Label FormikTextInput__LabelAsPlaceholder"
                  for="participantEmail0"
                >
                  Email
                </label>
                <input
                  class="FormikTextInput__Input"
                  id="participantEmail0"
                  name="participantEmail0"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div>
      <div
        style="margin-left: -15px; margin-right: -15px; display: flex; flex-wrap: wrap; flex-grow: 0; flex-shrink: 0; align-items: normal; justify-content: flex-start;"
      >
        <div
          style="box-sizing: border-box; min-height: 1px; position: relative; padding-left: 15px; padding-right: 15px; width: 100%; flex-basis: 0px; flex-grow: 1; flex-shrink: 0; max-width: 100%; margin-left: 0%;"
        >
          <div
            class="ParticipantForm__Participant"
          >
            <div
              class="ParticipantForm__Icon"
              style="background-color: rgb(255, 203, 23);"
            >
              <svg
                viewBox="0 0 512 512"
                width="25"
              >
                <path
                  class="prefix__active-path"
                  d="M421.522 360.98c-41.359 0-75.01 33.652-75.01 75.01 0 41.359 33.652 76.01 75.01 76.01 41.359 0 75.01-34.652 75.01-76.01 0-41.359-33.651-75.01-75.01-75.01zM405.011 62.517C365.338 22.844 311.596 0 255.5 0S146.662 22.844 106.989 62.517s-61.517 93.415-61.517 149.511v58.939h331.045v-75.01c0-8.292 6.71-15.002 15.002-15.002s15.002 6.71 15.002 15.002v136.534c4.943-.716 9.865-1.516 15.002-1.516 16.167 0 31.311 3.977 45.006 10.533v-129.48c-.001-56.096-21.845-109.837-61.518-149.511z"
                  data-old_color="#000000"
                  data-original="#000000"
                  fill="#fff"
                />
                <path
                  class="prefix__active-path"
                  d="M45.472 300.972c-16.571 0-30.004 13.433-30.004 30.004v30.004c0 16.571 13.433 30.004 30.004 30.004h281.57a105.212 105.212 0 0149.475-49.475v-40.537H45.472z"
                  data-old_color="#000000"
                  data-original="#000000"
                  fill="#fff"
                />
              </svg>
            </div>
            <p
              class="Text Text__h5"
            >
              Participant 2
            </p>
          </div>
        </div>
      </div>
      <div
        style="margin-left: -15px; margin-right: -15px; display: flex; flex-wrap: wrap; flex-grow: 0; flex-shrink: 0; align-items: normal; justify-content: flex-start;"
      >
        <div
          class="ParticipantForm__InputRow"
        >
          <div
            style="box-sizing: border-box; min-height: 1px; position: relative; padding-left: 15px; padding-right: 15px; width: 50%; flex-basis: 50%; flex-grow: 0; flex-shrink: 0; max-width: 50%; margin-left: 0%;"
          >
            <div
              class="ParticipantForm__Input"
            >
              <div
                class="FormikTextInput__Container"
              >
                <label
                  class="FormikTextInput__Label FormikTextInput__LabelAsPlaceholder"
                  for="participantName1"
                >
                  Name
                </label>
                <input
                  class="FormikTextInput__Input"
                  id="participantName1"
                  name="participantName1"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
          <div
            style="box-sizing: border-box; min-height: 1px; position: relative; padding-left: 15px; padding-right: 15px; width: 50%; flex-basis: 50%; flex-grow: 0; flex-shrink: 0; max-width: 50%; margin-left: 0%;"
          >
            <div
              class="ParticipantForm__Input"
            >
              <div
                class="FormikTextInput__Container"
              >
                <label
                  class="FormikTextInput__Label FormikTextInput__LabelAsPlaceholder"
                  for="participantEmail1"
                >
                  Email
                </label>
                <input
                  class="FormikTextInput__Input"
                  id="participantEmail1"
                  name="participantEmail1"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div>
      <div
        style="margin-left: -15px; margin-right: -15px; display: flex; flex-wrap: wrap; flex-grow: 0; flex-shrink: 0; align-items: normal; justify-content: flex-start;"
      >
        <div
          style="box-sizing: border-box; min-height: 1px; position: relative; padding-left: 15px; padding-right: 15px; width: 100%; flex-basis: 0px; flex-grow: 1; flex-shrink: 0; max-width: 100%; margin-left: 0%;"
        >
          <div
            class="ParticipantForm__Participant"
          >
            <div
              class="ParticipantForm__Icon"
              style="background-color: rgb(252, 182, 0);"
            >
              <svg
                viewBox="0 0 512 512"
                width="25"
              >
                <path
                  class="prefix__active-path"
                  d="M421.522 360.98c-41.359 0-75.01 33.652-75.01 75.01 0 41.359 33.652 76.01 75.01 76.01 41.359 0 75.01-34.652 75.01-76.01 0-41.359-33.651-75.01-75.01-75.01zM405.011 62.517C365.338 22.844 311.596 0 255.5 0S146.662 22.844 106.989 62.517s-61.517 93.415-61.517 149.511v58.939h331.045v-75.01c0-8.292 6.71-15.002 15.002-15.002s15.002 6.71 15.002 15.002v136.534c4.943-.716 9.865-1.516 15.002-1.516 16.167 0 31.311 3.977 45.006 10.533v-129.48c-.001-56.096-21.845-109.837-61.518-149.511z"
                  data-old_color="#000000"
                  data-original="#000000"
                  fill="#fff"
                />
                <path
                  class="prefix__active-path"
                  d="M45.472 300.972c-16.571 0-30.004 13.433-30.004 30.004v30.004c0 16.571 13.433 30.004 30.004 30.004h281.57a105.212 105.212 0 0149.475-49.475v-40.537H45.472z"
                  data-old_color="#000000"
                  data-original="#000000"
                  fill="#fff"
                />
              </svg>
            </div>
            <p
              class="Text Text__h5"
            >
              Participant 3
            </p>
          </div>
        </div>
      </div>
      <div
        style="margin-left: -15px; margin-right: -15px; display: flex; flex-wrap: wrap; flex-grow: 0; flex-shrink: 0; align-items: normal; justify-content: flex-start;"
      >
        <div
          class="ParticipantForm__InputRow"
        >
          <div
            style="box-sizing: border-box; min-height: 1px; position: relative; padding-left: 15px; padding-right: 15px; width: 50%; flex-basis: 50%; flex-grow: 0; flex-shrink: 0; max-width: 50%; margin-left: 0%;"
          >
            <div
              class="ParticipantForm__Input"
            >
              <div
                class="FormikTextInput__Container"
              >
                <label
                  class="FormikTextInput__Label FormikTextInput__LabelAsPlaceholder"
                  for="participantName2"
                >
                  Name
                </label>
                <input
                  class="FormikTextInput__Input"
                  id="participantName2"
                  name="participantName2"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
          <div
            style="box-sizing: border-box; min-height: 1px; position: relative; padding-left: 15px; padding-right: 15px; width: 50%; flex-basis: 50%; flex-grow: 0; flex-shrink: 0; max-width: 50%; margin-left: 0%;"
          >
            <div
              class="ParticipantForm__Input"
            >
              <div
                class="FormikTextInput__Container"
              >
                <label
                  class="FormikTextInput__Label FormikTextInput__LabelAsPlaceholder"
                  for="participantEmail2"
                >
                  Email
                </label>
                <input
                  class="FormikTextInput__Input"
                  id="participantEmail2"
                  name="participantEmail2"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </form>
</DocumentFragment>
`;
